pallene/constant_propagation.lua

```diff
@@ -0,0 +1,131 @@
+local ir = require "pallene.ir"
+
+local constant_propagation = {}
+
+local function is_constant_value(v)
+    local tag = v._tag
+    if     tag == "ir.Value.Nil"      then return true
+    elseif tag == "ir.Value.Bool"     then return true
+    elseif tag == "ir.Value.Integer"  then return true
+    elseif tag == "ir.Value.Float"    then return true
+    elseif tag == "ir.Value.String"   then return true
+    elseif tag == "ir.Value.LocalVar" then return false
+    elseif tag == "ir.Value.Function" then return true
+    else
+        error("impossible")
+    end
+end
+
+-- Replaces toplevel constant variables by their respetive values.
+--
+-- Currently assumes that the toplevel constant variable is initialized with a
+-- constant literal. Does not currently recognize non-trivial constant
+-- expressions as being constant.
+function constant_propagation.run(module)
+
+    local n_globals = #module.globals
+
+    -- 1) Find what toplevel variables are initialized to a constant in $init
+
+    local constant_initializer = {} -- { g_id => ir.Value? }
+    for i = 1, n_globals do
+        constant_initializer[i] = false
+    end
+
+    do
+        -- DFS traversal to find SetGlobal instructions with a constant
+        -- initializer. We ignore the instructions inside If, Loop, and For
+        -- statements, since those might be skipped.
+        local stack = { module.functions[1].body }
+        while #stack > 0 do
+            local cmd = table.remove(stack)
+            local tag = cmd._tag
+            if     tag == 'ir.Cmd.SetGlobal' then
+                if is_constant_value(cmd.src) then
+                    constant_initializer[cmd.global_id] = cmd.src
+                end
+            elseif tag == 'ir.Cmd.Seq' then
+                for i = #cmd.cmds, 1, -1 do
+                    table.insert(stack, cmd.cmds[i])
+                end
+            else
+                -- skip
+            end
+        end
+    end
+
+    -- 2) Find what toplevel variables are never re-initialized or never used
+
+    local n_reads  = {} -- { g_id => int }
+    local n_writes = {} -- { g_id => int }
+    for i = 1, n_globals do
+        n_reads[i]  = 0
+        n_writes[i] = 0
+    end
+
+    for _, func in ipairs(module.functions) do
+        for _, cmd in ipairs(ir.flatten_cmd(func.body)) do
```

Comment:
It would be nicer if we had an iterator for all commands instead of creating a new list and than using ipairs.
We could even keep this implementation, but hide it under a function.
```
for _, cmd in ir.flatten(func.body) do
```

Reply:
Good idea. There is now an ir.iter function replacing all uses of ir.flatten_cmd

I didn't delete the flatten_cmd because it is still used in one of my unpublished branches.